File: 162.js
// LeetCode 162 Find Peak element
// Reference: https://leetcode.com/problems/find-peak-element/solutions/692986/javascript-simple-binary-and-linear-search/?envType=featured-list&envId=top-interview-questions?envType=featured-list&envId=top-interview-questions

/**
 * @param {number[]} nums
 * @return {number}
 */
var findPeakElement = function(nums) {
    // Binary Search 
    // Time O(logN)
    let left = 0, right = nums.length-1, mid;

    while (left < right){
        mid = Math.floor((right + left)/2);
        if(nums[mid] > nums[mid+1]) right = mid;
        else left = mid + 1;
    }
    return left;
    
};

var findPeakElement2 = function(nums){
    // Linear Search
    // Time O(N)
    for(let i = 0; i < nums.length; i++){
        if(nums[i] > nums[i+1]) return i;
    }
    return nums.length-1;

}

File: 162.py
# LeetCode 162 Find Peak element
# Reference: https://www.youtube.com/watch?v=kMzJy9es7Hc

class Solution:
    def findPeakElement(self, nums: List[int]) -> int:

        # O(logN): Binary Search
        # Need to modify a bit
        l, r = 0, len(nums) - 1

        while l <= r:
            m = l + ((r-1)//2) # avoid overflowing

            # left neighbor greater
            if m > 0 and nums[m] < nums[m-1]:
                r = m - 1
            # right neighbor greater
            elif m < len(nums) - 1 and nums[m] < nums[m+1]:
                l = m + 1
            else:
                return m
        